[NVPTX] Add patterns for fma.relu.{f16|f16x2|bf16|bf16x2}

hdelan opened the pull request:
Add patterns to lower `fmaxnum(fma(a, b, c), 0)` to `fma.rn{.ftz}.relu` for `f16`, `f16x2`, `bf16`, `bf16x2` types, when `nnan` is used.

`fma_relu` honours `NaN`, so the substitution is only made if the `fma` is `nnan`, since `fmaxnum` returns the non NaN argument when passed a NaN value.

This patch also removes some `bf16` ftz instructions since `FTZ` is not supported with the `bf16` type, according to the PTX ISA docs.

Commit:
Add patterns for fma.relu.{f16|bf16}

Add patterns to lower fma(a, b, c) > 0 ? fma(a, b, c) : 0 for f16 and
bf16 types.

hdelan commented:
Ping @ldrumm @frasercrmck

justinfargnoli commented on llvm/lib/Target/NVPTX/NVPTXInstrInfo.td:
```diff
@@ -3917,3 +3917,22 @@ def atomic_thread_fence_seq_cst_cta :
+
+def FMARELU_F16 :
+  NVPTXInst<(outs Int16Regs:$dst), (ins Int16Regs:$a, Int16Regs:$b, Int16Regs:$c),
+            "fma.rn.relu.f16 \t$dst, $a, $b, $c;", []>;
```
Do we need the `Requires<...>` on the instruction too?

hdelan replied:
It's only being used by the anonymous patterns below, which have the necessary `Requires`. I don't think we need to introduce extra noise by repeating them here.

Artem-B replied:
I think applying constraint to the instruction itself is the right thing to do. We do not want them to be emitted unintentionally, even if we do not do it now.

I do not know whether the constraint propagates to the pattern, but I think it may, so applying it here should do the job. It's easy enough to test by running the tests while targeting an older GPU.

hdelan replied:
I've added the PTX and arch requirements on the instruction, and the pattern Requires just on the pattern.

justinfargnoli reviewed (approved):
Overall, LGTM! 

Please wait for @AlexMaclean's review though as he's more familiar with `NVPTXInstrInfo.td` than I am.

Artem-B commented on llvm/test/CodeGen/NVPTX/fma-relu.ll:
```diff
@@ -0,0 +1,77 @@
+; CHECK-NEXT:    fma.rn.relu %rs4, %rs1, %rs2, %rs3;
+; CHECK-NEXT:    st.param.b16 [func_retval0], %rs4;
+; CHECK-NEXT:    ret;
+  %1 = fmul half %a, %b
+  %2 = fadd half %1, %c
```
It would be good to add a couple more test runs:
 - one with w/o mul/add -> fma contraction to make sure we do not use `fma.rn.relu` unintentionally.
 - one targeting older GPUs to make sure we do not emit `fma.rn.relu` there.

hdelan replied:
I've added more tests to cover these cases.

Artem-B commented on llvm/lib/Target/NVPTX/NVPTXInstrInfo.td:
```diff
@@ -3917,3 +3917,22 @@ def atomic_thread_fence_seq_cst_cta :
+
+def : Pat<(f16 (fmaxnum (fma Int16Regs:$a, Int16Regs:$b, Int16Regs:$c), fpimm0)),
+  (FMARELU_F16 Int16Regs:$a, Int16Regs:$b, Int16Regs:$c)>,
+  Requires<[useFP16Math, allowFMA, allowUnsafeFPMath, hasPTX<60>, hasSM<70>]>;
+
+def : Pat<(bf16 (fmaxnum (fma Int16Regs:$a, Int16Regs:$b, Int16Regs:$c), fpimm0)),
+  (FMARELU_BF16 Int16Regs:$a, Int16Regs:$b, Int16Regs:$c)>,
+  Requires<[hasBF16Math, allowFMA, allowUnsafeFPMath, hasPTX<60>, hasSM<70>]>;
```
Constraints look wrong. According to https://docs.nvidia.com/cuda/parallel-thread-execution/index.html#half-precision-floating-point-instructions-fma

```
PTX ISA Notes
fma.relu.{f16, f16x2} and fma{.relu}.{bf16, bf16x2} introduced in PTX ISA version 7.0.
Target ISA Notes
fma.relu.{f16, f16x2} and fma{.relu}.{bf16, bf16x2} require sm_80 or higher.
```

hdelan replied:
Oops good catch. Thanks

hdelan replied:
Fixed

Artem-B commented on llvm/lib/Target/NVPTX/NVPTXInstrInfo.td:
```diff
@@ -3917,3 +3917,24 @@ def atomic_thread_fence_seq_cst_cta :
+
+def FMARELU_F16 :
+  NVPTXInst<(outs Int16Regs:$dst), (ins Int16Regs:$a, Int16Regs:$b, Int16Regs:$c),
+    "fma.rn.relu.f16 \t$dst, $a, $b, $c;", []>,
```
Next question is -- what do we want to do about `.ftz`?
We handle it for regular FMA instructions and it's probably needed here, too.

ldrumm commented on llvm/lib/Target/NVPTX/NVPTXInstrInfo.td:
```diff
@@ -3917,3 +3917,40 @@ def atomic_thread_fence_seq_cst_cta :
   NVPTXInst<(outs), (ins), "fence.acq_rel.cta;", []>,
   Requires<[hasPTX<60>, hasSM<70>]>;
+
+def fpimm0 : FPImmLeaf<fAny, [{
```
It's not clear from the name that this is strictly *positive* zero. Maybe `def positive_zero_fp` or a better equivalent if you can

hdelan replied:
Thanks have updated.

ldrumm reviewed (approved):
Looks good!

Artem-B commented on llvm/lib/Target/NVPTX/NVPTXInstrInfo.td:
```diff
@@ -3917,3 +3917,40 @@ def atomic_thread_fence_seq_cst_cta :
+    Requires<[hasBF16Math, hasPTX<70>, hasSM<80>]>;
+def FMARELU_F16_FTZ :
+  NVPTXInst<(outs Int16Regs:$dst), (ins Int16Regs:$a, Int16Regs:$b, Int16Regs:$c),
+    "fma.rn.relu.ftz.f16 \t$dst, $a, $b, $c;", []>,
```
The `relu` and `ftz` parts of the instructions appear to be in the order different from that in the PTX manual:

```
Syntax

fma.rnd{.ftz}.relu.f16      d, a, b, c;
fma.rnd{.ftz}.relu.f16x2    d, a, b, c;
```
Does ptxas accept generated instruction? In any case to avoid unnecessary divergence vs. the manual, it may be worth it to match the official syntax, even if ptxas happens to accept `fma.rn.relu.ftz.f16`

hdelan replied:
PTXAS did accept the instruction but I've updated to match the ordering specified here.

Artem-B reviewed (commented):
We will eventually want to figure out a way to effectively use f16x2 and bf16x2 variants of `fma.rn.relu`.

Commit:
Make sure FMA only has one use, update tests

FMA relu should only be emitted if the FMA node has a single use. This
should limit register pressure in some cases do avoid computing FMA as
well as FMA.relu.

Also split tests into two files, one using FMA contraction and the other
using the FMA intrinsic.

hdelan commented:
> Suppose the `fma` has more uses in addition to the `fmaxnum`, If this optimization kicks in it may increase the register pressure and won't be a clear win in terms of performance. I'm not sure this will be a problem, but to be conservative it may be better to implement this as a DAG combine and verify the `fma` has a single use.

I've changed the pattern matching to make sure FMA relu is only emitted if the FMA DAG has a single use.

hdelan commented:
> We will eventually want to figure out a way to effectively use f16x2 and bf16x2 variants of `fma.rn.relu`.

These should be available via the LLVM intrinsics. I think pattern matching to generate the x2 variants is a bit messy and fragile since it involves unpacking vector types. There also isn't really a concise way to express this pattern in source code, meaning the use cases for this sort of substitution are likely to be rare. Users needing this instruction for x2 types can always do so via the clang builtins.

Commit:
Rename fpimm0 to fpimm_positive_zero

Commit:
Add support for f16x2, bf16x2 PTX instructions

hdelan commented on llvm/test/CodeGen/NVPTX/fma-relu.ll:
```diff
@@ -0,0 +1,349 @@
+; NOTE: Assertions have been autogenerated by utils/update_llc_test_checks.py UTC_ARGS: --version 5
```
This prevents `llc` from emitting PTX in some cases so I'm not going to add this flag.

hdelan commented on llvm/test/CodeGen/NVPTX/fma-relu.ll:
```diff
@@ -0,0 +1,356 @@
+; RUN: %if ptxas %{ llc < %s -march=nvptx64 -mcpu=sm_80 -mattr=ptx70 -verify-machineinstrs -fp-contract=fast -nvptx-fma-level=2 | %ptxas-verify -arch=sm_80 %}
+; RUN: llc < %s -march=nvptx64 --enable-unsafe-fp-math -mcpu=sm_80 -mattr=ptx70 -verify-machineinstrs -fp-contract=fast -nvptx-fma-level=0 | FileCheck %s --check-prefixes=CHECK-NO-FMA
+; RUN: llc < %s -march=nvptx64 --enable-unsafe-fp-math -mcpu=sm_70 -mattr=ptx70 -verify-machineinstrs -fp-contract=fast -nvptx-fma-level=2 | FileCheck %s --check-prefixes=CHECK-NO-ARCH
+; RUN: llc < %s -march=nvptx64 --enable-unsafe-fp-math -mcpu=sm_70 -mattr=ptx70 -verify-machineinstrs -fp-contract=fast -nvptx-fma-level=2 | FileCheck %s --check-prefixes=CHECK-NO-PTX
```
Thanks have changed.

Commit:
Remove support for fma.rn.ftz{.relu}.bf16

The PTX ISA docs note that FTZ is not permitted for FTZ for bf16 types.

hdelan commented:
FTZ FMA is not supported for bf16, so I have removed some patterns elsewhere that were allowing `fma.rn.ftz.bf16` to be emitted.

Commit:
Update comment

fma.ftz isn't supported for bf16.

Artem-B commented on llvm/test/CodeGen/NVPTX/fma-relu-contract.ll:
```diff
@@ -0,0 +1,1422 @@
+; CHECK-SM70-NEXT:    setp.nan.f32 %p4, %f12, %f12;
+; CHECK-SM70-NEXT:    or.b32 %r40, %r36, 4194304;
+; CHECK-SM70-NEXT:    selp.b32 %r41, %r40, %r39, %p4;
+; CHECK-SM70-NEXT:    { .reg .b16 tmp; mov.b32 {tmp, %rs15}, %r41; }
+; CHECK-SM70-NEXT:    mov.b32 %r42, {%rs15, %rs13};
```
Another future optimization opportunity -- we could lower those "truncate-to-i16, then recombine into i32" as a single `prmt`

Artem-B commented on llvm/lib/Target/NVPTX/NVPTXInstrInfo.td:
```diff
@@ -3917,3 +3903,63 @@ def atomic_thread_fence_seq_cst_cta :
+    Requires<[useFP16Math, hasPTX<70>, hasSM<80>]>;
+
+// FTZ variants are only supported by fp16, not bf16
+def : Pat<(f16 (fmaxnum (NVPTX_fma_oneuse Int16Regs:$a, Int16Regs:$b, Int16Regs:$c), fpimm_positive_zero)),
```
How does FMA handle NaNs? `fmaxnum` requires `If either operand is a NaN, returns the other non-NaN operand.`

If we pass a NaN into an `fma.relu`, and get back a `NaN` back and that would be incorrect for `fmaxnum`.
The documentation does not seem to say anything about NaN handling by FMA (it only mentions it for .sat variants)

hdelan replied:
We are only applying this pattern if `allowUnsafeFPMath` is enabled. So AFAIA we don't need to worry about NaN propagation.

Artem-B replied:
The comment for the `allowUnsafeFPMath` says that it only affects precision of the operations, not NaN handling:
https://github.com/llvm/llvm-project/blob/7f60f1312ae007d645fc96618db1238baea088b9/llvm/include/llvm/Target/TargetOptions.h#L171
So, it would allow using a regular FMA instead of `a*b+c`, but `NaN` propagation would not be affected.
This is not the case for `fmaxnum` -> `fma`, where `fmaxnum` has distinctly different requirements for processing NaN inputs.

In any case, we should run that by someone who's familiar with FP handling nuances. 
@arsenm @RKSimon -- what would be a sufficient constraint to reflect that the instruction is not expected to deal with NaNs?

arsenm replied:
Definitely do not rely on allowUnsafeFPMath for no nans. We really should delete allowUnsafeFPMath. 

FMA propagates nan as normal, and may also produce output nans based on the operands.

I also think this pattern is broken for signed zeroes. I assume the ftz modifier is expected to preserve -0

hdelan replied:
Thanks, I have changed the constraints to `Requires<[noNaNsFPMath]>`.

> I also think this pattern is broken for signed zeroes. I assume the ftz modifier is expected to preserve -0

I have extended the fpimm pattern to also accommodate negative zero.

hdelan replied:
> I also think this pattern is broken for signed zeroes. I assume the ftz modifier is expected to preserve -0

I think I misunderstood what you were saying first time round. You are concerned with a comparison of `0.0` and `-0.0`. Since I am pattern matching with `fmaxnum`, which says either -0.0 or 0.0 may be returned in a comparison of both values, the guarantees of `fma.relu` (which doesn't specify what is returned when signed and unsigned zeros are in the final comparison) are the same as the `fmaxnum` comparison.

arsenm replied:
We are fixing the definition of fmaxnum/fminnum to require correctly ordered signed zeroes. The fuzzy behavior can be achieved with nsz, so it doesn't make sense to define it this way.

Best to respect this now to reduce the amount of code that needs changing

hdelan replied:
I've added a PatFrag to make sure the `fmaxnum` has `nsz`.

Commit:
Check for noNaNs instead of unsafeFPMath

The pattern substitution for fmax(fma() 0) is only valid if NaNs cannot
be emitted. This is because fmax is guaranteed to return the non-NaN
arg, whereas fma.relu is NaN preserving.

Commit:
Make fpimm match negative zeros as well

Extend the pattern to allow negative zeros for scalar types.

Artem-B commented on llvm/lib/Target/NVPTX/NVPTXInstrInfo.td:
```diff
@@ -3917,3 +3904,63 @@ def atomic_thread_fence_seq_cst_cta :
   NVPTXInst<(outs), (ins), "fence.acq_rel.cta;", []>,
   Requires<[hasPTX<60>, hasSM<70>]>;
+
+def fpimm_pos_neg_zero : FPImmLeaf<fAny, [{
```
`fpimm_any_zero` ?

hdelan replied:
Sure, have changed. Merging now.

Artem-B reviewed (approved):
Nice. Please hold off on merging the patch till Monday, so if we find issues, there's someone around to deal with that.

Commit:
Rename pos_neg_zero to any_zero

Commit:
Allow instruction flags to emit fma_relu

If nnan is used in instruction flags then FMA relu can also be emitted.

hdelan commented:
I've also added extra cases so that `nnan .. fma` in the same pattern will allow the emission of `fma.relu`

frasercrmck commented on llvm/lib/Target/NVPTX/NVPTXInstrInfo.td:
```diff
@@ -3917,3 +3904,86 @@ def atomic_thread_fence_seq_cst_cta :
+def : Pat<(v2bf16 (fmaxnum (NVPTX_fma_oneuse_and_nnan Int32Regs:$a, Int32Regs:$b, Int32Regs:$c), fpimm_positive_zero_v2bf16)),
+  (FMARELU_BF16X2 Int32Regs:$a, Int32Regs:$b, Int32Regs:$c)>;
+
+// Use function attributes for noNaNsFPMath, instead of instruction flag
```
I personally wouldn't bother supporting the function attribute case - I don't think it warrants the extra code and patterns. Supporting `nnan` should be enough. I think that the NVPTX backend should generally be moving towards using instruction flags rather than function attributes to do its various fast-math optimizations. But I'd defer to other people on that point.

frasercrmck commented on llvm/lib/Target/NVPTX/NVPTXInstrInfo.td:
```diff
@@ -3917,3 +3904,86 @@ def atomic_thread_fence_seq_cst_cta :
+  return N->hasOneUse() && N->getFlags().hasNoNaNs();
+}]>;
+
+def FMARELU_F16 :
```
I'm sure a `multiclass` could whittle this bulk down quite a bit. The only things changing are the instruction suffix, the reg class, and the instruction suffix in the asm.

frasercrmck reviewed (commented):
The PR description needs updating to reflect some of the new changes - the conditions under which the optimization is legal/performed, and that some instructions are being removed.

Commit:
Refactor instruction defs

Use a multiclass to refactor instruction defs.

Commit:
Check for func attrs in PatFrag

Instead of adding a flag for noNaNsFPMath, just add the check in the
PatFrag which also checks for single use FMA, as well as instruction
flags.

Commit:
Update comments

Artem-B commented on llvm/lib/Target/NVPTX/NVPTXInstrInfo.td:
```diff
@@ -3917,3 +3903,52 @@ def atomic_thread_fence_seq_cst_cta :
+    Requires<[hasBF16Math, hasPTX<70>, hasSM<80>]>;
+}
+
+defm FMARELU : FMARELU<Int16Regs>;
+defm FMARELU : FMARELU<Int32Regs, "x2">;
```
Nit: I'd prefer constructing the name up-front here, rather than deriving the final record name from the argument string.

E.g:
```
multiclass FMARELU<RegisterClass RC, string OpVecString = "">  {
  def _F16 : ...
  // FTZ variants are only supported by fp16, not bf16
  def _F16_FTZ : ...
  def _BF16: ...
}

defm FMARELU : FMARELU<Int16Regs>;
defm FMARELUv2 : FMARELU<Int32Regs, "x2">;
```

frasercrmck replied:
Sorry, that was my suggestion. I saw that `f16x2` and `bf16x2` were the more common names for these types in the instruction definitions, and thought it better to keep the instruction enums consistent at the expense of the TableGen. I can't see any existing examples of `v2_{type}`. There are a few examples of `v2{type}` but mostly for `b(8|16|32|64)` - not so much with specific types. In summary I suppose it's just a little inconsistent?

Artem-B replied:
I see. My concern was mostly this seemingly duplicate `defm FMARELU`, which is what got me to stop and poke at it. 

So, how can we express it all better. Ideally we want to have just one point to parametrize the multiclass.

We have 3 instruction variants, two vector sizes. From that we derive register size.
Considering that we're talking only about 6 items, there's not much point to do anything particularly complicated, as it would not improve anything over just enumerating things.

Actually, that may be the most concise way to do it -- just define a simple class for one instruction record that can parametrize ftz/type/regtype, and then define each of the 6 variants we have. You can see an example of this approach here:
https://github.com/llvm/llvm-project/blob/cb9481dbf902adc349757eca12a0a09396dc4a23/llvm/lib/Target/NVPTX/NVPTXInstrInfo.td#L1222

(side-note: FNEF_F16 and FNEG_BF16 could be further collapsed into a single class)

I think the net result will be shorter than what we have now.

hdelan replied:
I've updated to use `class` instead of `multiclass`

frasercrmck commented on llvm/lib/Target/NVPTX/NVPTXInstrInfo.td:
```diff
@@ -3917,3 +3903,52 @@ def atomic_thread_fence_seq_cst_cta :
   Requires<[hasPTX<60>, hasSM<70>]>;
+
+def fpimm_any_zero : FPImmLeaf<fAny, [{
+  return Imm.isExactlyValue(+0.0) | Imm.isExactlyValue(-0.0);
```
This might give a warning about bitwise orbeing used instead of logical or. I'd suggest using `||`.

Alternatively, perhaps `Imm.isExactlyValue(+0.0) | Imm.isExactlyValue(-0.0)` is just `Imm.isZero()`?

hdelan replied:
Have updated

Commit:
Use imm.isZero instead of isExactly

Commit:
Only use pattern when nsz also used

fmaxnum currently returns either -0.0 or 0.0 when comparing both values,
however this specification is going to change to return positive zero
in a comparison between both vals (according to arsenm).

fma.relu is not specified to return either positive or negative zero in
a comparison of signed zeros, so it shouldn't be emitted in a DAG that
uses fmaxnum's new specification unless nsz is also used.

Commit:
Use class for NVPTXInstr defs

Don't use multiclass. Class is more readable.

Commit:
Update tests after rebase

Artem-B commented on llvm/lib/Target/NVPTX/NVPTXInstrInfo.td:
```diff
@@ -3959,3 +3945,54 @@ def atomic_thread_fence_seq_cst_cta :
+  return N->getFlags().hasNoSignedZeros() || TM.Options.NoSignedZerosFPMath;
+}]>;
+
+class FMARELU<RegisterClass RC, string OpVecString, list<Predicate> Preds>
```
Nit: `OpVecString` should probably be called `Instruction` now.

Side note: It looks like we've simplified it to a pretty generic `3-operand instruction` helper class. We could probably use it in other places. It seems to be a fairly common pattern in other back-ends.

hdelan replied:
Thanks. Have updated. Since the commits are going to be squashed here I will leave the refactoring for a subsequent patch.

Commit:
Rename helper class

TODO: refactor other methods to use this.
